### PR TITLE
feat: Added TopicName and SubsciptionName resolution from container as an extension point.

### PR DESCRIPTION
**Pull request by maw136**

### CODE REVIEW

**Comment by maw136** [comment]
https://github.com/googleapis/google-cloud-dotnet/issues/12213 better aproach

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2023 Google LLC
+// Copyright 2024 Google LLC
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs:1]: Please revert this change - we leave copyright years from when the file was first introduced, where possible. (It's harder for generated code, which is generated from scratch each time.)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -35,12 +35,29 @@ public static partial class ServiceCollectionExtensions
...
+    /// </summary>
+    /// <param name="services">
+    /// The <see cref="IServiceCollection"/> to add the singleton client to.
+    /// </param>
+    /// <param name="action">
+    /// An action delegate to invoke on the <see cref="PublisherClientBuilder"/> for configuring the <see cref="PublisherClient"/>. This is invoked before <paramref name="services"/> are used.
+    /// Must not be null and at-least <see cref="PublisherClientBuilder.TopicName"/> must be set.
+    /// </param>
+    /// <returns>The updated <see cref="IServiceCollection"/>, for method chaining.</returns>
+    public static IServiceCollection AddPublisherClient(this IServiceCollection services, Action<IServiceProvider, PublisherClientBuilder> action)
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs:54]: This is *strictly speaking* a breaking change due to an existing call becoming ambiguous, in the following scenarios:

- Using a null literal - I'm not bothered about that, as we're validating that the action is non-null anyway (so it would only break something at compile-time that would already be broken at execution time)
- Using a method group conversion where there are two overloads: I think this is really, really unlikely, and the benefits of this outweigh the risk

**Review by jskeet — CHANGES_REQUESTED** [review]
I *think* this will be okay - I'll discuss the breaking change aspect with Amanda.
Separately, I'll fix up whitespace that was spotted in the other PR.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (maw136) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs:1]: Will do. I haven't changed it manually. There may be some build time script or task that updates it.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs (hunk 2 above)
Comment (maw136) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs:54]: Ah, the sweet harmony of shared concerns. If you're nodding in agreement, I'm all in.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (maw136) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/ServiceCollectionExtensions.cs:1]: Yes, I have some minimal reformat rules on save enabled. .editorconfig is probably one of them rules sources.
